lib/ts/recipe/totp/types.ts

```diff
@@ -0,0 +1,143 @@
+/* Copyright (c) 2023, VRAI Labs and/or its affiliates. All rights reserved.
+ *
+ * This software is licensed under the Apache License, Version 2.0 (the
+ * "License") as published by the Apache Software Foundation.
+ *
+ * You may not use this file except in compliance with the License. You may
+ * obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+import OverrideableBuilder from "supertokens-js-override";
+import { BaseRequest, BaseResponse } from "../../framework";
+import { NormalisedAppinfo } from "../../types";
+import { SessionContainer } from "../session";
+
+export type TypeInput = {
+    issuer: string;
+    defaultSkew?: number;
+    defaultPeriod?: number;
+    allowUnverifiedDevice?: boolean;
```

Comment:
what is the use case of making this true for the user?

Reply:
Actually adding up with MFA, I think this can go wrong.

Original effect of this was that if a user scans the QR but doesn't verify at the time of TOTP registration, they will still be prompted for a TOTP code in the next login (since `devices.length > 0` mean isFactorEnabledForUser gives `True`)

But if you think about it, if user doesn't scan the QR and closes the tab, they will be still be prompted for a TOTP code in the next login attempt. But now since they never scanned it, they don't have any way to enter a valid TOTP. 

There's no way to differentiate these two cases. So I think we should remove this param altogether or just set it to False.

Reply:
@KShivendu missed this?

Reply:
> Will the listDevices return 0 or 1 result? If it returns 0, and if the user creates a new device, what will happen to the older one? How will that be cleaned up?

As per the current implementation, it will return 1 if device has been created. So user will asked to enter TOTP while logging in next time. We can make listDevices so that, it also takes allowUnverifiedDevices param (or we can filter by verified status on the SDK level). But for the device cleanup part, user needs to remove the unused device on their own. We can automate this based on expiry maybe if we don't allow any unverified devices for logging in.

Reply:
I don't want to change the core. What do other providers do in this case - a user adds a device, but does not verify it, and closes the browser. Then the user signs in - do they see a totp screen? Or do they have to setup a new device?

Reply:
Most providers don't allow more than one device. Only AWS does. So we used DevTools and found that in AWS devices are created when you pass a device name but those devices remain "unassigned" (equivalent to verified in our case)

These unverfied devices weren't getting deleted (atleast not instantly) and aren't shown to the users. If user creates a device with the same name, frontend calls delete device API to remove the existing one and create a new one.

Reply:
- I think we should just remove the config for `allowUnverifiedDevice` from the core alltogether (since we are anyway changing the core now).
- If devices are unverified in the core, they remain as unverified. If a new device is created with the same name as the unverified device, the core should just replace the existing unverified device with the new one.
- Finally, move the name generation strategy to the core - so the createDevice API in the core should take an optional device name, and the name will be determined based on the number of verified devices. This also means that we no longer need to have the name generation logic on the backend SDK level. Keep in mind that the core can still return device already exists if the name is given by the user, and that name belongs to a verified device already.